### PR TITLE
Refactor structured transforms

### PR DESCRIPTION
**Pull request by Tratcher**
This takes the work from #612 and applies it to structured transforms in preparation for #60 and #416. It incorporates feedback about the transforms being too structured, unable to arbitrarily manipulate the request/response, rename headers, do async work, etc.

The original design:
- Distinguished between request and request header transforms.
- Applied transforms while copying request/response headers. This was done for efficiency but required a complex indexing system and limited you to one transform rule per header name.
- Transforms are sync
- Shared transform logic for response headers and trailers, individual transforms only look at header values passed to them.

The new design:
- Unified request and request header transforms.
- Headers are copied before transforms are applied (if enabled). Transforms are then applied to the result. This is a bit less efficient since you need to copy the value, remove it, and then replace it, but it only affects headers with transforms, not every header. It's also a bit more work to write transforms against HttpRequestMessage/HttpContent but I've added some helper methods to smooth this out (TakeHeader, AddHeader).
- Transforms are more powerful now, they can affect the whole request rather than just a specific header/field.
- Transforms are async
- Split response header and trailer transforms since transforms now need to deal with the source and destination of their data.

I know this is a lot of churn. The most interesting changes to review are:
- [TransformBuilder](https://github.com/microsoft/reverse-proxy/pull/646/files#diff-0fb89f2699f9c7d3143b84021d8d836b2fcd1b1ec0a0223a28bc432d2b3592f8)
- [StructuredTransformer](https://github.com/microsoft/reverse-proxy/pull/646/files#diff-113396a09a3cd8e1fa7f4c65d14d79c3b7804dc84f8188f356dcf108da29fee8)
- [RequestTransform](https://github.com/microsoft/reverse-proxy/pull/646/files#diff-2e296d77a68dc839ce72d7838e78ddeb2eeb2d07ee7de1892a61e06b56c95e10)
- [ResponseTransform](https://github.com/microsoft/reverse-proxy/pull/646/files#diff-c3965c2ef1f5d65223fe80dd76f549d2c549d807f6bfdafdafc7d0ca58b8dc10)

### CODE REVIEW

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransformContext.cs [hunk 1]
```diff
@@ -1,29 +1,34 @@
...
         /// applying header transforms.
         /// </summary>
-        public HttpRequestMessage ProxyRequest { get; internal set; }
+        public HttpRequestMessage ProxyRequest { get; init; }
+
+        /// <summary>
+        /// Gets or sets if the request headers have been copied from the HttpRequest to the HttpRequestMessage and HttpContent.
+        /// Transforms use this when searching for the current value of a header they should operate on.
+        /// </summary>
+        public bool HeadersCopied { get; set; }
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransformContext.cs:31]: Why isn't this also `init`? I can't think of a scenario if a transform would want to change this property.

**Review by Kahbazi — APPROVED** [review]

File: src/ReverseProxy/Service/RuntimeModel/Transforms/StructuredTransformer.cs [hunk 2]
```diff
@@ -20,192 +20,137 @@ namespace Microsoft.ReverseProxy.Service.RuntimeModel.Transforms
...
-            Dictionary<string, RequestHeaderTransform> requestHeaderTransforms,
-            Dictionary<string, ResponseHeaderTransform> responseHeaderTransforms,
-            Dictionary<string, ResponseHeaderTransform> responseTrailerTransforms)
+        internal StructuredTransformer(bool? copyRequestHeaders, bool? copyResponseHeaders, bool? copyResponseTrailers,
+            IList<RequestTransform> requestTransforms,
+            IList<ResponseTransform> responseTransforms,
+            IList<ResponseTrailersTransform> responseTrailerTransforms)
         {
             ShouldCopyRequestHeaders = copyRequestHeaders;
+            ShouldCopyResponseHeaders = copyResponseHeaders;
```
Comment (Kahbazi) [inline src/ReverseProxy/Service/RuntimeModel/Transforms/StructuredTransformer.cs:32]: We could set the value in constructor only once and change the `bool?` to `bool`.
```suggestion
            ShouldCopyResponseHeaders = copyResponseHeaders.GetValueOrDefault(true);
```

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransformContext.cs (hunk 1 above)
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransformContext.cs:31]: We worked with a customer recently that was building their own proxy with Kestrel and HttpClient and they ran into header encoding issues in both directions. Such issues could be dealt with here by disabling the default copy logic and providing a transform that handled the encoding issues while copying the headers. That transform would set HeadersCopied when it was done so subsequent transforms would know where to look for headers.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        /// then the HttpContent, and falling back to the HttpContext only if
+        /// <see cref="RequestTransformContext.HeadersCopied"/> is not set.
+        /// This ordering allows multiple transforms to mutate the same header.
+        /// </summary>
+        /// <param name="headerName">The name of the header to take.</param>
+        /// <returns>The requested header value, or StringValues.Empty if none.</returns>
+        public static StringValues TakeHeader(RequestTransformContext context, string headerName)
+        {
+            var existingValues = StringValues.Empty;
+            if (context.ProxyRequest.Headers.TryGetValues(headerName, out var values))
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs:32]: I don't expect that to be a problem. The standard approach for adding a header is to try adding it to the HttpRequestMessage and fall back to the HttpContent if that fails. Even if that were reversed there shouldn't be duplicate headers. These HttpRequest/ResponseMessages aren't created by arbitrary user code, they're created by HttpProxy or HttpClient so the way they add headers should be predictable.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /// <summary>
+    /// The base class for request transforms.
+    /// </summary>
+    public abstract class RequestTransform
+    {
+        /// <summary>
+        /// Transforms any of the available fields before building the outgoing request.
+        /// </summary>
+        public abstract Task ApplyAsync(RequestTransformContext context);
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs:19]: I tried that but ValueTask is not well supported on 3.1. I'm not sure it matters as much for Tasks without T results, we're able to return the same CompletedTask when it's not async.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <param name="headerName">The name of the header to take.</param>
+        /// <returns>The requested header value, or StringValues.Empty if none.</returns>
+        public static StringValues TakeHeader(RequestTransformContext context, string headerName)
+        {
+            var existingValues = StringValues.Empty;
+            if (context.ProxyRequest.Headers.TryGetValues(headerName, out var values))
+            {
+                context.ProxyRequest.Headers.Remove(headerName);
+                existingValues = values.ToArray();
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs:35]: I'll try casting for now and see if there are cases later that break it.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/StructuredTransformer.cs (hunk 2 above)
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/StructuredTransformer.cs:32]: I'd left it this way so I could test TransformBuilder for the cases where no config was provided vs an explicit negative config given. If it shows up in a trace then I'll remove it and find another way to test the builder.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+        /// <summary>
+        /// Removes and returns the current header value by first checking the HttpRequestMessage,
+        /// then the HttpContent, and falling back to the HttpContext only if
+        /// <see cref="RequestTransformContext.HeadersCopied"/> is not set.
+        /// This ordering allows multiple transforms to mutate the same header.
+        /// </summary>
+        /// <param name="headerName">The name of the header to take.</param>
+        /// <returns>The requested header value, or StringValues.Empty if none.</returns>
+        public static StringValues TakeHeader(RequestTransformContext context, string headerName)
```
Comment (Tratcher) [inline src/ReverseProxy/Service/RuntimeModel/Transforms/RequestTransform.cs:29]: Would these Take and Add methods make more sense as instance methods on the RequestTransformContext?

**Review by alnikola — APPROVED** [review]
LGTM

**Review by davidfowl — APPROVED** [review]

**Pull request merged**